docs/tutorials/auction.md

```diff
@@ -0,0 +1,244 @@
+---
+sidebar_position: 4
+---
+
+# Tutorial 4: Auction
+
+## Overview
+
+In this tutorial, we will go over how to build an auction contract. It is open and transparent, where everyone can participate and the highest bidder wins when the bidding is over.
+
+There are two ways to interact with the contract:
+
+1. Bid: if a higher bid is found, the current highest bidder is updated, and the previous highest bidder is refunded.
+2. Close: the auctioneer can close the auction after it expires and take the offer.
+
+## Contract Properties
+
+According to the interactions above, this contract needs to store three properties:
+
+- The auctioneer, who starts the auction
+- The deadline for the auction
+- The highest bidder until now
+
+```ts
+// The bidder's address.
+@prop(true)
+bidder: PubKeyHash
+
+// The auctioneer's public key.
+@prop()
+readonly auctioneer: PubKey
+
+// Deadline of the auction. Can be block height or timestamp.
+@prop()
+readonly auctionDeadline: bigint
+```
+
+## Constructor
+
+Initialize all the `@prop` properties in the constructor. Note that we don't need to pass a `bidder` parameter.
+
+```ts
+constructor(auctioneer: PubKey, auctionDeadline: bigint) {
+    super(...arguments)
+    // the initial bidder is the auctioneer himeself
+    this.bidder = hash160(auctioneer)
+    this.auctioneer = auctioneer
+    this.auctionDeadline = auctionDeadline
+}
+```
+
+When deploying the contract, the auctioneer locked the minimal bid into the contract, and at this time, the highest bidder would be himself.
+
+```ts
+const auction = new Auction(publicKeyAuctioneer, auctionDeadline)
+const deployTx = await auction.deploy(minBid)
+```
+
+## Public Methods
+
+### Bid
+
+In method `public bid(bidder: PubKeyHash, bid: bigint)`, we need to check if the bidder has a higher bid than the previous one. If so, we update the highest bidder in the contract state and refund the previous bidder.
+
+We can read the previous highest bid from the balance of the contract UTXO.
+
+```ts
+const highestBid: bigint = this.ctx.utxo.value
+```
+
+Then it's easy to demand a higher bid.
+
+```ts
+assert(bid > highestBid, 'the auction bid is lower than the current highest bid')
+```
+
+The spending/redeeming tx has these outputs.
```

Comment:
Add a tx diagram here, similar to https://plutus-pioneer-program.readthedocs.io/en/latest/pioneer/week1.html#the-english-auction-contract